Feature/pybind11 build

elhuhdron opened the pull request:
This pull request is so that changes in feature/pybind11 still can build standalone in Windows and with gcc in linux. I tested builds with:

- Windows 10, Anaconda 3.7 2019.07, Visual Studio 2017
- macOS, Anaconda 3.7 2019.07, clang 7.3
- Ubuntu 18.04, Anaconda 3.7 2019.07, gcc 7.4 and clang 6.0

The changes to the source / headers were only to maintain msvc and gcc support.

I made a few changes to setup.py:
- For me, without the `packages = ['pylibczi'],` argument to `setup(`, none of the python classes install into site-packages
- I'd like to maintain using a pip-installed cmake if it's available, which for me makes the whole build a bit more self-contained (particularly in Windows). I improved the code that searches for this cmake in setup.py. The default is still just to use whatever cmake is in the path (as before). I'm very happy that most of the other build details have been pushed into cmake!
- In order to still allow building with gcc on linux, I modified setup to use the env variables if they are already set, but still default to clang, so for example build with `CC=gcc CXX=g++ pip install .[all]`
- For build debug, I find it very useful to have the verbose makefile mode turned on, so I tied this to the `DISTUTILS_DEBUG` existing env variable.

heeler commented:
I've incorporated pieces of this into master and with @toloudis managed to track down some code that caused weird MSVC compiler behavior and fix feature/pybind11 so it builds on all platforms. 

I just added gcc to the compiler matrix so it triggers both clang++ and g++. 
It build on [osx, ubuntu] x [g++, clang++] and (windows w/ MSVC)

The setup.py changes look useful would you prefer I pull them into feature/pybind11 or rebase feature/pybind11_build off of feature/pybind11 and then resolve the conflicts manually? I appreciate your help with this. 

For your main repo if you're interested in using github actions it would probably be good to sign up for the beta at https://github.com/features/actions. Then when we eventually pull back to your home then we can just setup the secrets and the build testing and such should all work.

elhuhdron commented:
> I've incorporated pieces of this into master and with @toloudis managed to track down some code that caused weird MSVC compiler behavior and fix feature/pybind11 so it builds on all platforms.
> I just added gcc to the compiler matrix so it triggers both clang++ and g++.
> It build on [osx, ubuntu] x [g++, clang++] and (windows w/ MSVC)

Fantastic!

> The setup.py changes look useful would you prefer I pull them into feature/pybind11 or rebase feature/pybind11_build off of feature/pybind11 and then resolve the conflicts manually? I appreciate your help with this.

I can rebase, no problem. I'll try to get to it tomorrow, but I'll let you know.

> For your main repo if you're interested in using github actions it would probably be good to sign up for the beta at https://github.com/features/actions. Then when we eventually pull back to your home then we can just setup the secrets and the build testing and such should all work.

Sounds good. I signed up; I'm unfamiliar, so I'll read up on the documentation when I get a chance.

heeler commented:
I had a lot of good feedback on my naming conventions so I'm going through the code and applying it just to make it more readable. If you hold off on rebasing till I'm finished that would probably be best. I'm hoping to be done in a few hours.

heeler commented:
I finished refactoring the names and such. You should be good to rebase. Any future changes should be very minor.

elhuhdron commented:
> I finished refactoring the names and such. You should be good to rebase. Any future changes should be very minor.

Ok, all set! The new head of pybind11 is building perfect for me, and cmake chooses gcc automatically now (however you fixed that :)

Rebase is complete, only changes on top of the branch now are: 
- utilizing the pip-installed cmake if available
- verbose makefile mode tied to DISTUTILS_DEBUG existing env variable

heeler commented on setup.py:
```diff
@@ -91,11 +93,42 @@ def __init__(self, name, sourcedir=''):
+                        valid_cmake = False
+                    if valid_cmake and (not found_cmake or cmake_depth < found_cmake_depth):
+                        found_cmake = cmake_exe
+                        found_cmake_depth = cmake_depth
```
I'm a little concerned that the code seems to be overriding the convention of using the first occurrence in $PATH with using the "shallowest" occurrence in $PATH. The user can reasonably shuffle PATH order but if they have to read the setup.py to figure out how to get it to build that seems non-ideal. The intent of looping through till you find a working one seems reasonable but maybe test the version number of output is >= 2.8.12 (the minimum requirement in the CMakeLists.txt file so that it doesn't pass the setup.py and then blow up in `cmake ..`.

elhuhdron replied:
> I'm a little concerned that the code seems to be overriding the convention of using the first occurrence in $PATH with using the "shallowest" occurrence in $PATH.

Yeah, that's a good point. Before the last set of build changes you introduced, the build command was not taking the pip-installed cmake. I thought this was even after the addition of the cmake package as a build requirement, but now I can not reproduce this. I tested again and it's indeed taking pip-installed cmake on both platforms without this change.

So, I agree this could potentially override the user's path and additionally seems no longer necessary.

The other change is super minor, so I'll delete this pull request.

elhuhdron commented:
pybind11 branch is now building properly on all tested platforms, so pull request is no longer necessary.
